gradio-app/gradio
# Implement JS Client tests

## Description

Adds tests for `@gradio/client`

- [x]  `client`
- [x] `duplicate`
- [x] `init_helpers.ts` (`resolve_root`, `get_jwt`, `map_names_to_ids`, `determine_protocol`) 
- [x] `data.ts` (`walk_and_store_blobs`, `update_object`, `skip_queue`, `post_message`)
- [x]  `api_info.ts` (`handle_message`, `get_description`, `get_type`, `process_endpoint`)
- [x]  `view_api.ts` 
- [x] `spaces.ts` (`set_space_timeout`, `get_space_hardware`, `check_space_status`, `discussions_enabled`)
- [x] `upload_files.ts`
- [x] `post_data.ts`


Tests new and deprecated syntax: 

- `await Client.connect()`
- deprecated `await client()`
- `Client.duplicate()`
- deprecated `duplicate`
- Client class overrides

To test separately: 

- [ ] `stream.ts`
- [ ] `submit.ts`
- [ ] `predict.ts`

I'd like to write tests in a separate PR for these because 1) there are lots of tests to write for `submit()` 2) testing SSE events is a little complicated (but doable) 3) I'd also like to see if we can test with a local spun up demo as we do in the Python client in that PR

Implicitly tested and could do with unit tests:

- `resolve_config`
- `handle_blob.ts`


Closes: #4658 

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`
